File: Hashmap/Group Anagrams/GroupAnagrams.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"sort"
	"strconv"
	"strings"
)

func groupAnagrams(strs []string) [][]string {
	anagramMap := make(map[string][]string)
	for _, str := range strs {
		sortedStr := sortString(str)
		anagramMap[sortedStr] = append(anagramMap[sortedStr], str)
	}
	var result [][]string
	for _, group := range anagramMap {
		result = append(result, group)
	}
	return result
}

func sortString(s string) string {
	runes := []rune(s)
	sort.Slice(runes, func(i, j int) bool {
		return runes[i] < runes[j]
	})
	return string(runes)
}

func main() {
	reader := bufio.NewReader(os.Stdin)
	tcStr, _ := reader.ReadString('\n')
	tc, _ := strconv.Atoi(strings.TrimSpace(tcStr))
	for t := 0; t < tc; t++ {
		strsStr, _ := reader.ReadString('\n')
		strs := strings.Fields(strings.TrimSpace(strsStr))
		result := groupAnagrams(strs)
		for _, group := range result {
			fmt.Println(strings.Join(group, " "))
		}
		fmt.Println()
	}
}

File: Hashmap/Group Anagrams/GroupAnagrams.cpp
#include <algorithm>
#include <iostream>
#include <string>
#include <unordered_map>
#include <vector>
using namespace std;

class Solution {
 public:
  vector<vector<string>> groupAnagrams(vector<string>& strs) {
    unordered_map<string, vector<string>> anagramGroups;
    for (const string& s : strs) {
      string sortedStr = s;
      sort(sortedStr.begin(), sortedStr.end());
      anagramGroups[sortedStr].push_back(s);
    }
    vector<vector<string>> result;
    for (const auto& group : anagramGroups) {
      result.push_back(group.second);
    }
    return result;
  }
};

int main() {
  Solution obj;
  int tc;
  cin >> tc;
  while (tc-- > 0) {
    int n;
    cin >> n;
    vector<string> strs(n);
    for (int i = 0; i < n; i++) {
      cin >> strs[i];
    }
    vector<vector<string>> result = obj.groupAnagrams(strs);
    cout << "Anagram groups:\n";
    for (const auto& group : result) {
      for (const string& word : group) {
        cout << word << " ";
      }
      cout << endl;
    }
    cout << endl;
  }
  return 0;
}
